### PR TITLE
[Clang][Sema] Use correct TemplateName when transforming TemplateSpecializationType

### LINKED ISSUE #91677: Crash on missing `template` disambiguator after C++20-allowed missing `typename` disambiguator
https://godbolt.org/z/shvqo8jhY
```
template <typename> struct t1 {
  template <typename>
  struct t2 {};
};

template <typename T>
t1<T>::t2<T> f1();

void f2() {
  f1<bool>();
}
```

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-16.0.0/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -S --gcc-toolchain=/opt/compiler-explorer/gcc-12.2.0 -fcolor-diagnostics -fno-crash-diagnostics -std=c++20 <source>
1.	<eof> parser at end of file
2.	<source>:9:6: LLVM IR generation of declaration 'f2'
3.	<source>:9:6: Generating code for declaration 'f2'
 #0 0x0000569be3cb48df llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x3ba18df)
 #1 0x0000569be3cb290c llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x3b9f90c)
 #2 0x0000569be3c05518 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007599e2a42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x0000569be357cac5 llvm::FunctionType::get(llvm::Type*, llvm::ArrayRef<llvm::Type*>, bool) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x3469ac5)
 #5 0x0000569be4449873 clang::CodeGen::CodeGenTypes::GetFunctionType(clang::CodeGen::CGFunctionInfo const&) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4336873)
 #6 0x0000569be423c394 clang::CodeGen::CodeGenTypes::ConvertFunctionTypeInternal(clang::QualType) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4129394)
 #7 0x0000569be423d7f8 clang::CodeGen::CodeGenTypes::ConvertType(clang::QualType) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x412a7f8)
 #8 0x0000569be41fd6bf clang::CodeGen::CodeGenModule::GetAddrOfFunction(clang::GlobalDecl, llvm::Type*, bool, bool, clang::CodeGen::ForDefinition_t) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40ea6bf)
 #9 0x0000569be44a558c EmitFunctionDeclPointer(clang::CodeGen::CodeGenModule&, clang::GlobalDecl) CGExpr.cpp:0:0
#10 0x0000569be44b405a EmitDirectCallee(clang::CodeGen::CodeGenFunction&, clang::GlobalDecl) CGExpr.cpp:0:0
#11 0x0000569be44cd525 clang::CodeGen::CodeGenFunction::EmitCallee(clang::Expr const*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x43ba525)
#12 0x0000569be44cd94a clang::CodeGen::CodeGenFunction::EmitCallExpr(clang::CallExpr const*, clang::CodeGen::ReturnValueSlot) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x43ba94a)
#13 0x0000569be4516727 (anonymous namespace)::ScalarExprEmitter::VisitCallExpr(clang::CallExpr const*) CGExprScalar.cpp:0:0
#14 0x0000569be450b5e0 (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) CGExprScalar.cpp:0:0
#15 0x0000569be450da97 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x43faa97)
#16 0x0000569be44aebb6 clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x439bbb6)
#17 0x0000569be44cbce9 clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x43b8ce9)
#18 0x0000569be4135593 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4022593)
#19 0x0000569be413b7aa clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40287aa)
#20 0x0000569be419b08d clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x408808d)
#21 0x0000569be41a62a6 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40932a6)
#22 0x0000569be41fd88a clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40ea88a)
#23 0x0000569be41fa015 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40e7015)
#24 0x0000569be41fa503 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x40e7503)
#25 0x0000569be4201ff9 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.part.0) CodeGenModule.cpp:0:0
#26 0x0000569be4fa6d01 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) ModuleBuilder.cpp:0:0
#27 0x0000569be4f9973d clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4e8673d)
#28 0x0000569be61c9a74 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x60b6a74)
#29 0x0000569be4fa4865 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4e91865)
#30 0x0000569be4880671 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x476d671)
#31 0x0000569be4804993 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x46f1993)
#32 0x0000569be4964b3b clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x4851b3b)
#33 0x0000569be148d41c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x137a41c)
#34 0x0000569be1488bdc ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#35 0x0000569be4665ce9 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#36 0x0000569be3c05977 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x3af2977)
#37 0x0000569be4665f1c clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#38 0x0000569be462e3fc clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x451b3fc)
#39 0x0000569be462eded clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x451bded)
#40 0x0000569be463923c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x452623c)
#41 0x0000569be148b523 clang_main(int, char**) (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x1378523)
#42 0x00007599e2a29d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#43 0x00007599e2a29e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#44 0x0000569be1484d1e _start (/opt/compiler-explorer/clang-16.0.0/bin/clang+++0x1371d1e)
```

### PR DESCRIPTION
**Pull request by jcsxky**
Consider the following testcase:
```cpp
namespace PR12884_original {
  template <typename T> struct A {
    struct B { ##1
      template <typename U> struct X {};
      typedef int arg;
    };
    struct C {
      typedef B::X<typename B::arg> x; 
    };
  };

  template <> struct A<int>::B { ##2
    template <int N> struct X {};
    static const int arg = 0;
  };

  A<int>::C::x a;
}
```
It will crash when compiling with `clang(assertions trunk)`. The reason is that we lookup `X`(`B::X`) in `##1` when instantiating `typedef B::X<typename B::arg> x; ` during instantiating `A<int>::C::x`. This is incorrect because we should lookup `X` in `##2` when we see the declaration `A<int>::C::x a;`. Since clang parse `A<T>::B<T>` to an `ElaboratedType`(`typename` is not required while compiling with `-std=c++20`)  while `typename A<T>::B<T>` turns to be a `DependentTemplateSpecializationType`, we should rebuild the `TemplateName` with transformed `Qualifier`(whose type is `NestedNameSpecifier`) to make sure the lookup context is correct.
This patch also attempts to fix #91677 which crashes with the same reason.

### CODE REVIEW

File: clang/test/SemaTemplate/typename-specifier-3.cpp [hunk 1]
```diff
@@ -28,16 +28,17 @@ namespace PR12884_original {
       typedef int arg;
     };
     struct C {
-      typedef B::X<typename B::arg> x; // precxx17-warning{{missing 'typename' prior to dependent type name B::X; implicit 'typename' is a C++20 extension}}
```
Comment (zyn0217) [inline clang/test/SemaTemplate/typename-specifier-3.cpp:31]: A drive-by comment: The test `PR12884_original` is still crashing on trunk but it slips through our testing probably because we somehow stop after errors above - if we move the whole namespace to the top, then we'd immediately run into an assert saying `Unable to find instantiation of declaration!`.

File: clang/lib/Sema/TreeTransform.h [hunk 2]
```diff
@@ -7216,9 +7216,30 @@ TreeTransform<Derived>::TransformElaboratedType(TypeLocBuilder &TLB,
...
+    CXXScopeSpec SS;
+    SS.Adopt(QualifierLoc);
+    if (TemplateDecl *TD = TST->getTemplateName().getAsTemplateDecl()) {
+      TemplateName InstName = getDerived().RebuildTemplateName(
+          SS, TL.getTemplateKeywordLoc(), *TD->getIdentifier(),
+          TL.getNamedTypeLoc().getBeginLoc(), /*ObjectType=*/QualType(),
+          /*FirstQualifierInScope=*/nullptr, /*AllowInjectedClassName=*/false);
+      if (InstName.isNull())
+        return QualType();
+      NamedT = TransformTemplateSpecializationType(TLB, SpecTL, InstName);
```
Comment (zyn0217) [inline clang/lib/Sema/TreeTransform.h:7235]: Did you mean `getDerived().TransformTemplateSpecializationType()`?

File: clang/lib/Sema/TreeTransform.h [hunk 3]
```diff
@@ -7216,9 +7216,30 @@ TreeTransform<Derived>::TransformElaboratedType(TypeLocBuilder &TLB,
...
-  QualType NamedT = getDerived().TransformType(TLB, TL.getNamedTypeLoc());
-  if (NamedT.isNull())
-    return QualType();
+  QualType NamedT;
+  if (SemaRef.getLangOpts().CPlusPlus20 && QualifierLoc &&
+      isa<TemplateSpecializationType>(TL.getNamedTypeLoc().getType())) {
+    TemplateSpecializationTypeLoc SpecTL =
+        TL.getNamedTypeLoc().castAs<TemplateSpecializationTypeLoc>();
+    const TemplateSpecializationType *TST =
+        SpecTL.getType()->castAs<TemplateSpecializationType>();
```
Comment (zyn0217) [inline clang/lib/Sema/TreeTransform.h:7225]:
```c++
if (QualType QT = TL.getNamedTypeLoc().getType(); ... && isa<...>(QT)) {
  auto SpecTL = ...;
  auto *TST = QT.getTypePtr();
}
```

Probably looks better.

**Review by mizvekov — COMMENTED** [review]
The problem here is the loss of the qualification on the TemplateNames

This patch fixes the problem, without taking any workarounds: https://github.com/llvm/llvm-project/pull/93433

It also doesn't cause any change in diagnostics in `clang/test/SemaTemplate/typename-specifier-3.cpp`.

**Comment by jcsxky** [comment]
> The problem here is the loss of the qualification on the TemplateNames
> 
> This patch fixes the problem, without taking any workarounds: #93433
> 
> It also doesn't cause any change in diagnostics in `clang/test/SemaTemplate/typename-specifier-3.cpp`.

I think we should report this message since `B::arg` in `A<int>` is a NTTP.

**Comment by mizvekov** [comment]
> > The problem here is the loss of the qualification on the TemplateNames
> > This patch fixes the problem, without taking any workarounds: #93433
> > It also doesn't cause any change in diagnostics in `clang/test/SemaTemplate/typename-specifier-3.cpp`.
> 
> I think we should report this message since `B::arg` in `A<int>` is a NTTP.

Yeah, I am not saying that is wrong, but this needs more investigation.

I am not looking at this bug specifically: I just noticed my patch would have an effect here, and after testing it sure did, but I haven't looked closely.

Maybe I used the term 'fixes' incorrectly, as in my patch makes it not crash, but it's strange that type sugar would have a semantic effect here, that is not supposed to happen.

Again, thinking in first principles, without looking at this example too specifically, my intuition says that what might be happening here is that we transformed a `DependentTemplateSpecializationType` into a `TemplateSpecializationType` too early, before resolving everything that needs to be resolved related to it's TemplateName and NestedNameSpecifier.

**Review by mizvekov — COMMENTED** [review]
So After we have formed a TemplateSpecializationType, we are done with the NNS. We keep it around in an ElaboratedType for type sugar only, it should not be needed to compile the program correctly anymore.

So I think this solution violates one important aspect of our AST design.

If we still need the NNS, we have to hold-off on building the TST until we don't.

**Comment by jcsxky** [comment]
> So After we have formed a TemplateSpecializationType, we are done with the NNS. We keep it around in an ElaboratedType for type sugar only, it should not be needed to compile the program correctly anymore.
> 
> So I think this solution violates one important aspect of our AST design.
> 
> If we still need the NNS, we have to hold-off on building the TST until we don't.

Ah, I see. Thanks for your explanation on the motivation of AST design!
